### ISSUE TITLE
Filter by Consultee

### ISSUE THREAD

**Issue by KyEggleston** [bcgov/foi-flow#4133]
* As a user of the document reviewer
* I want to filter the package in the document reviewer by specific consultees
* so that I can see the different pages/packages that may be provided to each consultee

**Assumptions & Scope**	
This story is to allow users to filter the package by specific consultees (i.e., the ones selected when applying the 'Consult' page flag). Rather than just selecting the 'Consult' page flag filter, and displaying all pages flagged with the 'Consult' page flag (and potentially numerous different consultees), users want the ability to filter by specific consultees that have been applied in the package. 

For example, in the image below, rather than only being able to filter by the 'Consult' page flag itself, you should instead be able to filter by the specific consultees under the 'Consult' page flag. In this example, [the design here](https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/985918996) would show AED, EMC and HSG as options to select under the 'Consult' page flag filter. Users will also have the ability to select all consultees, and unselect all consultees, from within the 'Consult' page flag filter dropdown.

![image](https://github.com/bcgov/foi-flow/assets/117312323/8df259ea-f3e6-4389-a3ce-a7fbe866ba18)

Only consultees that have been selected/added when applying the 'Consult' page flag should be available under the 'Consult' page flag filter. 

Users should be able to select one or more consultees under the 'Consult' page flag filter.

When a consultee is removed from all 'Consult' page flags used in the package, they should no longer appear under the 'Consult' page flag filter. 

This should work with the other filter criteria available (e.g., page flag filter, Organize By and Filter Records), combining to produce a broader, or more specific, search. 

What is IN scope?

Updating the filter options in the navigation (left) pane to also include consultees as filter options

What is NOT in scope?

Creating, saving or downloading a consult package
Creating a consult layer

**Acceptance Criteria**	

**Scenario 1: Consult Flag Filter - No Consultees**
* GIVEN a user is in the redaction app
* AND no pages have the 'Consult' page flag
* WHEN a user clicks the 'Consult' page flag filter
* THEN no consultees will be listed or selectable
* AND the user will be unable to filter by the 'Consult' page flag

**Scenario 2: Consult Flag Filter - 1 or more Consult Page Flags**
* GIVEN a user is in the redaction app
* AND =>1 page(s) have the 'Consult' page flag
* AND =>1 consultee is selected
* WHEN a user clicks the 'Consult' page flag filter
* THEN all consultees that have been selected in the 'Consultations' modal within the package will be listed or selectable from the dropdown (checkbox) (https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/975452421)

**Scenario 3: Select One Consultee**
* GIVEN a user is in the redaction app
* AND =>1 page(s) have the 'Consult' page flag
* AND =>1 consultee has been selected with the 'Consult' page flag
* WHEN a user clicks the 'Consult' page flag filter
* AND the user selects the checkbox for one consultee from the dropdown
* THEN the package will be filtered to display pages/files that have the consult page flag with that consultee selected 

**Scenario 4: Select Two or More Consultees**
* GIVEN a user is in the redaction app
* AND =>1 page(s) have the 'Consult' page flag
* AND =>2 Consultees have been selected with the 'Consult' page flag
* WHEN a user clicks the 'Consult' page flag filter
* AND the user selects the checkboxes for two or more consultees from the dropdown
* THEN the package will be filtered to display pages/files that have the consult page flag with at least one of the consultees selected 

**Scenario 5: Remove 'Consult' Page Flag Filter Consultees**
* GIVEN a user has already selected to filter down the package by a consultee
* WHEN a user clicks the 'Consult' page flag filter
* AND the user unselects the checkbox for the selected consultee
* THEN the package will no longer be filtered according to that consultee

**Scenario 6: Update / Remove Consultees from Filter**
* GIVEN a user has applied the 'Consult' page flag and consultee(s) to a page(s) in the package
* WHEN the user removes or modifies the consultee(s) in the package 
* THEN the 'Consult' page flag filter will update to reflect new, or removed, consultee(s)

**Scenario 7: Combine Page Flag Filters**
* GIVEN a user has filtered the package by the 'Consult' page flag and consultee(s)
* WHEN the user selects an additional page flag to filter by
* THEN the package will be filtered to show pages/files that have either one of the page flags

**Scenario 8: Select All Consultees**
* GIVEN a user has applied the 'Consult' page flag and consultee(s) to a page(s) in the package
* AND the user is in the 'Consult' page flag filter dropdown
* AND the 'Select / Unselect All' checkbox is empty
* WHEN the user clicks the 'Select / Unselect All' checkbox
* THEN all the consultees in the dropdown will be selected

**Scenario 9: Unselect All Consultees**
* GIVEN a user has applied the 'Consult' page flag and consultee(s) to a page(s) in the package
* AND the user is in the 'Consult' page flag filter dropdown
* AND the 'Select / Unselect All' checkbox is checked
* WHEN the user clicks the 'Select / Unselect All' checkbox
* THEN all the consultees in the dropdown will be unselected

**Dependencies? What is the impact of this dependency? (If so, link dependency in the ticket, make it visible in a team´s backlog)**

Blocked by #4221

**Validation Rules? (If yes, list here)**	

**Design**
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/975452421

**Definition of Ready**

1. [ ] Is there a well articulated User Story?
2. [ ] Is there Acceptance Criteria that covers all scenarios (happy/sad paths)?
3. [ ] If there is a user interface, is there a design? 
4. [ ] Does the user story need user research/validation?
5. [ ] Does this User Story needs stakeholder approval?
6. [ ] Design / Solution accepted by Product Owner
7. [ ] Is this user story small enough to be completed in a Sprint? Should it be split?
8. [ ] Are the dependencies known/ understood? (technical, business, regulatory/policy)
9. [ ] Has the story been estimated?

**Definition of Done**

1.    [ ] Passes developer unit tests
2.    [ ] Passes peer code review
3.    [ ] If there's a user interface, passes UX assurance
4.    [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
5.	[ ] Confirm Test cases built and succeeding 
6.	[ ] No regression test failures
7.	[ ] Test coverage acceptable by Product Owner
8.	[ ] Ticket ready to be merged to master or story branch
9.	[ ] Developer to list Config changes/ Update documents and designs
10.	[ ] Can be demoed in Sprint Review 
11.	[ ] Tagged as part of a Release
12.   [ ] Feature flagged if required
13.  [ ] Change Management activities done?

**Comment by KyEggleston** [comment]
@JHarrietha-AOT - here is the ticket for filtering down the package by consultee, and design can be added to it

**Comment by KyEggleston** [comment]
@m-prodan @lmullane @liseandtea @JHarrietha-AOT - updated based on designs, but will finish tomorrow morning - may need to confirm a few details

**Comment by KyEggleston** [comment]
Do we need a story to allow users to select all consults? Or do we think it's sufficient for them to select them individually?

**Comment by KyEggleston** [comment]
It's possible for this bug https://github.com/bcgov/foi-flow/issues/3562#issue-1646077480 to be looked at with this ticket, as it both pertains to filtering the package?

**Comment by m-prodan** [comment]
I was just thinking that Kyle. re: seeing 'all' consults. Maybe we just add an option for 'All' at the bottom?

**Comment by KyEggleston** [comment]
Or perhaps a 'Select / Unselect All' checkbox at the top of the list? That's a common UI approach to this

**Comment by JHarrietha-AOT** [comment]
@KyEggleston @m-prodan that makes sense - I would vote for a checkbox dropdown then (like watcher) and users can select whichever they need. 

I can updated the design to have checkboxes after Sprint Planning - Kyle are you okay to work on updating the ticket without the designs as it is just check boxes and I can add the updated link by EOD

**Comment by KyEggleston** [comment]
I can update the ticket - to confirm, are we adding the 'Select / Unselect All' checkbox to the top? Just will need to write ACs for it

**Comment by KyEggleston** [comment]
Updated ACs and will add design links after update

**Comment by JHarrietha-AOT** [comment]
@KyEggleston updated designs can be found here: https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/975452421

**Comment by m-prodan** [comment]
Note - will have a dependency to require standardizing page numbers between left panel and PDF tron - write ticket and estimate at next SP.

**Comment by KyEggleston** [comment]
@aparna-aot - this looks good. One observation:

When adding a consult/consultee page flag for the first time (e.g., HSG is a consultee on Page 1), it shows up in the Consult filter dropdown once (this is correct). 

![image](https://github.com/bcgov/foi-flow/assets/117312323/55a22859-b384-494f-8c50-fa35debebd63)

![image](https://github.com/bcgov/foi-flow/assets/117312323/aea4ddd3-0be2-4f06-babb-d7b36a123c27)

However, when adding that same consultee to another page, it duplicates the consultee in the Consult filter dropdown. In this example below, I also added HSG consultee/consult page flag to Page 3. As you can see, it duplicates the consultee in the Consult filter dropdown. 

![image](https://github.com/bcgov/foi-flow/assets/117312323/e94d0aa3-6805-4500-ade7-1b705a462e5e)

![image](https://github.com/bcgov/foi-flow/assets/117312323/5c125899-47d3-4dd2-9489-66c6dd0868d4)

This logic will continue on, for all subsequent pages flagged with the Consult page flag, with HSG as a consultee.

![image](https://github.com/bcgov/foi-flow/assets/117312323/76bd9277-4e2d-4373-9ec6-02f95e99d36a)
![image](https://github.com/bcgov/foi-flow/assets/117312323/224a63d9-fae1-4d97-b4f1-014e169dc4f4)

This wasn't captured in an AC or assumptions. Here's what an AC would look like for it:

GIVEN a consultee has has been selected on one page when applying a Consult page flag 
AND the consultee shows up once in the Consult filter dropdown
WHEN the same consultee is added to an additional page(s) when applying a Consult page flag 
THEN the Consult filter dropdown should still only display the consultee once
AND that consultee, when filtered, will show all pages flagged with that consultee
AND once all pages flagged with that consultee have the consultee removed, the consultee will no longer be displayed in the Consult filter dropdown

**Comment by aparna-aot** [comment]
@KyEggleston That definitely make sense. Even though the AC was not captured I have added the code for filter uniqueness. Let me quickly check the issue. Thanks for noticing that.

**Comment by aparna-aot** [comment]
@KyEggleston The fix is pushed to test-marshal now. You can retest it. Thanks!

**Comment by KyEggleston** [comment]
Thanks @aparna-aot - this works now. 

One other observation - when a user creates a custom public body for consult through the open text field in the Consult modal, if the character count exceeds 17 character (including spaces), the public body name is cut off in the Consult filter dropdown. 

![image](https://github.com/bcgov/foi-flow/assets/117312323/781f0fc7-db8d-4696-8a73-f537031e2b1d)

I created #4277, as this UI/formatting issue also occurs in the Consultations modal itself (albeit when the public body exceeds 28 characters, including spaces). 

I suppose the options are to wrap text or limit character count. Thoughts on wrapping text in Consult filter dropdown?

**Comment by aparna-aot** [comment]
@KyEggleston I can see multiple words are coming correctly may be the issue is with a single word with such length.
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBaUFQIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--8d092c5338ce06142a1ae160b7063d94c1cc3a9f/image.png)

**Comment by aparna-aot** [comment]
@KyEggleston Yeah. Issue is with a single word having this much length. Multiple words are wrapped correctly. Also technically will there be a single word with such length for consult name? Also if its captured in another ticket, we can work on it and close this ticket.
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBaUlQIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--1e8b35e37e6e46fbea03cf52ddec2a17ac79e43a/image.png)

**Comment by KyEggleston** [comment]
@aparna-aot - thanks for checking this. Very unlikely a single word will exceed this character limit, so I think this is fine. This ticket will be closed, and I will close the other ticket as well.

**Comment by KyEggleston** [comment]
All ACs pass, closing.
